File: ports/flutter/test/bundle_test.dart
// The precompiled bundle (docs/PROTOCOL.md 7.4): a namespace the manifest covers in the
// current language is seeded from the shipped file, with the bundle's cursor, instead of
// fetched, at boot and on a language switch. Storage wins only when newer and in the same
// language. Everything not covered keeps the fetch, and a miss still POSTs.
import 'dart:convert';

import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:i18n_keyless/i18n_keyless.dart';

const bundleCursor = '1757000000000';

final manifest = BundleManifest.fromJson({
  'primaryLanguage': 'fr',
  'languages': ['en', 'es', 'fr'],
  'exportedAt': bundleCursor,
  'namespaces': {
    'default': {
      'lastRefresh': bundleCursor,
      'languages': ['en', 'es', 'fr']
    },
    'checkout': {
      'lastRefresh': bundleCursor,
      'languages': ['en']
    },
  },
});

const files = <String, Map<String, String>>{
  'default/en.json': {'Bonjour': 'Hello', 'Merci': 'Thanks'},
  'default/es.json': {'Bonjour': 'Hola', 'Merci': 'Gracias'},
  'default/fr.json': {'Bonjour': 'Bonjour', 'Merci': 'Merci'},
  'checkout/en.json': {'Panier': 'Cart'},
};

class Loads {
  final List<String> calls = [];
  I18nKeylessBundle bundle({bool failing = false}) => I18nKeylessBundle(
        manifest: manifest,
        load: (namespace, lang) async {
          calls.add('$namespace/${lang.code}');
          if (failing) throw StateError('missing');
          return files['$namespace/${lang.code}.json'];
        },
      );
}

/// A backend that records requests and answers every dictionary from the API.
class Server {
  final List<http.Request> requests = [];
  List<http.Request> get dictionaries =>
      requests.where((r) => r.method == 'GET').toList();
  List<http.Request> get translates => requests
      .where((r) => r.method == 'POST' && r.url.path == '/translate')
      .toList();

  http.Client get client => MockClient((request) async {
        requests.add(request);
        http.Response json(Map<String, dynamic> body) =>
            http.Response(jsonEncode(body), 200,
                headers: {'content-type': 'application/json'});
        if (request.method == 'GET') {
          return json({
            'ok': true,
            'data': {
              'translations': {'Fetched': 'From the API'},
              'uniqueId': null,
              'lastRefresh': '111',
            },
            'error': '',
            'message': '',
          });
        }
        return json({'ok': true, 'data': {}, 'error': '', 'message': ''});
      });
}

I18nKeylessConfig configFor(
  Server server,
  MemoryStorage storage, {
  I18nKeylessBundle? bundle,
  LanguagesConfig? languages,
}) =>
    I18nKeylessConfig(
      apiKey: 'k',
      apiUrl: 'https://api.test',
      languages: languages ??
          const LanguagesConfig(
              primary: Lang.fr, supported: [Lang.fr, Lang.en, Lang.es]),
      storage: storage,
      bundle: bundle,
      httpClient: server.client,
      logger: (_) {},
    );

Future<MemoryStorage> storageWith(Map<String, String> entries) async {
  final storage = MemoryStorage();
  for (final entry in entries.entries) {
    await storage.setItem(entry.key, entry.value);
  }
  return storage;
}

void main() {
  group('boot', () {
    test(
        'seeds every bundled namespace in the current language, with the bundle cursor, and fetches nothing',
        () async {
      final server = Server();
      final storage = await storageWith({StorageKeys.currentLanguage: 'en'});
      final loads = Loads();
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: loads.bundle()));
      await client.waitForIdle();

      expect(client.currentLanguage, Lang.en);
      expect(client.getTranslation('Bonjour'), 'Hello');
      expect(client.getTranslation('Panier', namespace: 'checkout'), 'Cart');
      expect(client.translations,
          {'Bonjour': 'Hello', 'Merci': 'Thanks', 'Panier': 'Cart'});
      expect(server.dictionaries, isEmpty);
      expect(loads.calls, containsAll(['default/en', 'checkout/en']));
      // Persisted like a fetched dictionary, so the delta cursor survives a reload.
      expect(storage.entries[StorageKeys.lastRefresh], bundleCursor);
      expect(storage.entries[StorageKeys.lastRefreshKeyFor('checkout')],
          bundleCursor);
      expect(jsonDecode(storage.entries[StorageKeys.namespaces]!),
          ['default', 'checkout']);
    });

    test(
        'seeds the bundled primary dictionary too, and fetches only an origin namespace the bundle does not cover',
        () async {
      final server = Server();
      final storage = await storageWith({
        StorageKeys.currentLanguage: 'fr',
        StorageKeys.originNamespaces: jsonEncode(['chat']),
      });
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: Loads().bundle()));
      await client.waitForIdle();

      expect(jsonDecode(storage.entries[StorageKeys.translations]!),
          {'Bonjour': 'Bonjour', 'Merci': 'Merci'});
      expect(server.dictionaries.length, 1);
      expect(server.dictionaries.single.url.toString(),
          'https://api.test/translate/fr?last_refresh=null&namespace=chat');
    });

    test('fetches a language the bundle does not cover for a namespace',
        () async {
      final server = Server();
      final storage = await storageWith({StorageKeys.currentLanguage: 'es'});
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: Loads().bundle()));
      await client.waitForIdle();

      // default/es is bundled, checkout/es is not.
      expect(client.getTranslation('Bonjour'), 'Hola');
      expect(server.dictionaries.length, 1);
      expect(server.dictionaries.single.url.toString(),
          'https://api.test/translate/es?last_refresh=null&namespace=checkout');
      expect(
          jsonDecode(storage.entries[StorageKeys.translationsKeyFor('checkout')]!),
          {'Fetched': 'From the API'});
    });

    test(
        'keeps a stored slice on top of the bundle when it is newer and in the same language',
        () async {
      final server = Server();
      final newer = '${int.parse(bundleCursor) + 60000}';
      final storage = await storageWith({
        StorageKeys.currentLanguage: 'en',
        StorageKeys.namespaces: jsonEncode(['default']),
        StorageKeys.translations: jsonEncode({'Bonjour': 'Hello, reviewed'}),
        StorageKeys.lastRefresh: newer,
      });
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: Loads().bundle()));
      await client.waitForIdle();

      expect(client.getTranslation('Bonjour'), 'Hello, reviewed');
      expect(client.getTranslation('Merci'), 'Thanks');
      expect(storage.entries[StorageKeys.lastRefresh], newer);
      expect(server.dictionaries, isEmpty);
    });

    test('ignores a stored slice that is older than the bundle', () async {
      final server = Server();
      final storage = await storageWith({
        StorageKeys.currentLanguage: 'en',
        StorageKeys.namespaces: jsonEncode(['default']),
        StorageKeys.translations: jsonEncode({'Bonjour': 'Old hello'}),
        StorageKeys.lastRefresh: '${int.parse(bundleCursor) - 60000}',
      });
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: Loads().bundle()));
      await client.waitForIdle();

      expect(client.getTranslation('Bonjour'), 'Hello');
      expect(storage.entries[StorageKeys.lastRefresh], bundleCursor);
    });

    test('never mixes in a newer stored slice written in another language',
        () async {
      final server = Server();
      // Storage holds English, the app boots in Spanish (skipCurrentLanguageHydration).
      final storage = await storageWith({
        StorageKeys.currentLanguage: 'en',
        StorageKeys.namespaces: jsonEncode(['default']),
        StorageKeys.translations: jsonEncode({'Bonjour': 'Hello'}),
        StorageKeys.lastRefresh: '${int.parse(bundleCursor) + 60000}',
      });
      final client = I18nKeylessClient();
      await client.init(configFor(
        server,
        storage,
        bundle: Loads().bundle(),
        languages: const LanguagesConfig(
          primary: Lang.fr,
          supported: [Lang.fr, Lang.en, Lang.es],
          initWithDefault: Lang.es,
          skipCurrentLanguageHydration: true,
        ),
      ));
      await client.waitForIdle();

      expect(client.currentLanguage, Lang.es);
      expect(client.getTranslation('Bonjour'), 'Hola');
      expect(jsonDecode(storage.entries[StorageKeys.translations]!),
          {'Bonjour': 'Hola', 'Merci': 'Gracias'});
      expect(storage.entries[StorageKeys.lastRefresh], bundleCursor);
    });

    test('falls back to the fetch when the loader throws', () async {
      final server = Server();
      final storage = await storageWith({StorageKeys.currentLanguage: 'en'});
      final client = I18nKeylessClient();
      await client.init(
          configFor(server, storage, bundle: Loads().bundle(failing: true)));
      await client.waitForIdle();

      expect(server.dictionaries.length, 2);
      expect(client.getTranslation('Fetched'), 'From the API');
    });
  });

  group('language switch', () {
    test(
        'seeds the new language from the bundle and never mixes the previous language in',
        () async {
      final server = Server();
      final storage = await storageWith({StorageKeys.currentLanguage: 'en'});
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: Loads().bundle()));
      await client.waitForIdle();
      expect(server.dictionaries, isEmpty);

      await client.setCurrentLanguage(Lang.es);
      await client.waitForIdle();

      expect(client.getTranslation('Bonjour'), 'Hola');
      expect(jsonDecode(storage.entries[StorageKeys.translations]!),
          {'Bonjour': 'Hola', 'Merci': 'Gracias'});
      expect(storage.entries[StorageKeys.lastRefresh], bundleCursor);
      // checkout has no Spanish file: fetched.
      expect(server.dictionaries.length, 1);
      expect(server.dictionaries.single.url.toString(),
          'https://api.test/translate/es?last_refresh=null&namespace=checkout');
    });

    test('a miss still POSTs, and the delta after it starts from the bundle cursor',
        () async {
      final server = Server();
      final storage = await storageWith({StorageKeys.currentLanguage: 'en'});
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage, bundle: Loads().bundle()));
      await client.waitForIdle();

      expect(client.getTranslation('Nouveau'), 'Nouveau');
      await client.waitForIdle();

      expect(server.translates.length, 1);
      expect(jsonDecode(server.translates.single.body)['key'], 'Nouveau');
      expect(server.dictionaries.length, 1);
      expect(server.dictionaries.single.url.toString(),
          'https://api.test/translate/en?last_refresh=$bundleCursor');
      expect(client.getTranslation('Fetched'), 'From the API');
      expect(client.getTranslation('Bonjour'), 'Hello');
    });
  });

  group('without a bundle', () {
    test('fetches exactly as before', () async {
      final server = Server();
      final storage = await storageWith({StorageKeys.currentLanguage: 'en'});
      final client = I18nKeylessClient();
      await client.init(configFor(server, storage));
      await client.waitForIdle();

      expect(server.dictionaries.length, 1);
      expect(server.dictionaries.single.url.toString(),
          'https://api.test/translate/en?last_refresh=null');
    });
  });

  group('loadI18nKeylessBundleFromAssets', () {
    test('reads the manifest and the files through the asset bundle', () async {
      final assets = _FakeAssets({
        'assets/i18n-keyless/manifest.json': jsonEncode({
          'primaryLanguage': 'fr',
          'languages': ['en', 'fr'],
          'exportedAt': bundleCursor,
          'namespaces': {
            'default': {
              'lastRefresh': bundleCursor,
              'languages': ['en', 'fr']
            },
          },
        }),
        'assets/i18n-keyless/default/en.json': jsonEncode({'Bonjour': 'Hello'}),
      });
      final bundle = await loadI18nKeylessBundleFromAssets('assets/i18n-keyless/',
          assetBundle: assets);
      expect(bundle.manifest.primaryLanguage, 'fr');
      expect(bundle.manifest.namespaces['default']!.languages, ['en', 'fr']);
      expect(await bundle.load('default', Lang.en), {'Bonjour': 'Hello'});
      expect(bundleCovers(bundle.manifest, 'default', 'es'), isFalse);
    });
  });
}

class _FakeAssets extends CachingAssetBundle {
  _FakeAssets(this.files);

  final Map<String, String> files;

  @override
  Future<ByteData> load(String key) async {
    final text = files[key];
    if (text == null) throw StateError('asset not found: $key');
    return ByteData.sublistView(Uint8List.fromList(utf8.encode(text)));
  }
}

File: ports/kotlin/src/test/kotlin/io/i18nkeyless/BundleTest.kt
package io.i18nkeyless

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.util.concurrent.CopyOnWriteArrayList

/**
 * The precompiled bundle (docs/PROTOCOL.md 7.4): a namespace the manifest covers in the
 * current language is seeded from the shipped file, with the bundle's cursor, instead of
 * fetched, at boot and on a language switch. Storage wins only when newer and in the same
 * language. Everything not covered keeps the fetch, and a miss still POSTs.
 */
class BundleTest {
    companion object {
        const val CURSOR = "1757000000000"

        val manifest = BundleManifest.fromJson(
            mapOf(
                "primaryLanguage" to "fr",
                "languages" to listOf("en", "es", "fr"),
                "exportedAt" to CURSOR,
                "namespaces" to mapOf(
                    "default" to mapOf("lastRefresh" to CURSOR, "languages" to listOf("en", "es", "fr")),
                    "checkout" to mapOf("lastRefresh" to CURSOR, "languages" to listOf("en")),
                ),
            ),
        )

        val files = mapOf(
            "default/en" to mapOf("Bonjour" to "Hello", "Merci" to "Thanks"),
            "default/es" to mapOf("Bonjour" to "Hola", "Merci" to "Gracias"),
            "default/fr" to mapOf("Bonjour" to "Bonjour", "Merci" to "Merci"),
            "checkout/en" to mapOf("Panier" to "Cart"),
        )
    }

    /** A bundle whose loader records its calls. */
    class Loads(private val failing: Boolean = false) {
        val calls = CopyOnWriteArrayList<String>()
        val bundle = I18nKeylessBundle(manifest) { namespace, lang ->
            calls.add("$namespace/${lang.code}")
            if (failing) throw IllegalStateException("missing")
            files["$namespace/${lang.code}"]
        }
    }

    private fun storageWith(vararg entries: Pair<String, String>): MemoryStorage =
        MemoryStorage().apply { for ((key, value) in entries) setItem(key, value) }

    private fun boot(
        transport: FakeTransport,
        storage: MemoryStorage,
        bundle: I18nKeylessBundle?,
        languages: LanguagesConfig = LanguagesConfig(Lang.FR, listOf(Lang.FR, Lang.EN, Lang.ES)),
    ): I18nKeylessClient {
        val client = newClient(transport)
        client.init(
            I18nKeylessConfig(
                apiKey = "k-bundle",
                apiUrl = "https://api.test",
                languages = languages,
                storage = storage,
                bundle = bundle,
                logger = quietLogger(),
            ),
        ).get()
        client.waitForIdle()
        return client
    }

    private fun slice(storage: MemoryStorage, namespace: String = DEFAULT_NAMESPACE): Map<String, Any?>? =
        storage.getItem(StorageKeys.translationsKeyFor(namespace))?.let { Json.parse(it).asMap() }

    @Test
    fun `boot seeds every bundled namespace in the current language, with the bundle cursor, and fetches nothing`() {
        val transport = FakeTransport(mapOf("Fetched" to "From the API"))
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "en")
        val loads = Loads()
        val client = boot(transport, storage, loads.bundle)

        assertEquals(Lang.EN, client.currentLanguage)
        assertEquals("Hello", client.t("Bonjour"))
        assertEquals("Cart", client.t("Panier", namespace = "checkout"))
        assertEquals(mapOf("Bonjour" to "Hello", "Merci" to "Thanks", "Panier" to "Cart"), client.translationsSnapshot)
        assertTrue(transport.dictionaries.isEmpty())
        assertEquals(setOf("default/en", "checkout/en"), loads.calls.toSet())
        // Persisted like a fetched dictionary, so the delta cursor survives a reload.
        assertEquals(CURSOR, storage.getItem(StorageKeys.LAST_REFRESH))
        assertEquals(CURSOR, storage.getItem(StorageKeys.lastRefreshKeyFor("checkout")))
        assertEquals(listOf("default", "checkout"), Json.parse(storage.getItem(StorageKeys.NAMESPACES)!!))
    }

    @Test
    fun `boot seeds the bundled primary dictionary too, and fetches only an origin namespace the bundle does not cover`() {
        val transport = FakeTransport(mapOf("Fetched" to "From the API"))
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "fr", StorageKeys.ORIGIN_NAMESPACES to "[\"chat\"]")
        boot(transport, storage, Loads().bundle)

        assertEquals(mapOf("Bonjour" to "Bonjour", "Merci" to "Merci"), slice(storage))
        assertEquals(listOf("https://api.test/translate/fr?last_refresh=null&namespace=chat"), transport.dictionaries.map { it.url })
    }

    @Test
    fun `boot fetches a language the bundle does not cover for a namespace`() {
        val transport = FakeTransport(mapOf("Fetched" to "From the API"))
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "es")
        val client = boot(transport, storage, Loads().bundle)

        // default/es is bundled, checkout/es is not.
        assertEquals("Hola", client.t("Bonjour"))
        assertEquals(listOf("https://api.test/translate/es?last_refresh=null&namespace=checkout"), transport.dictionaries.map { it.url })
        assertEquals(mapOf("Fetched" to "From the API"), slice(storage, "checkout"))
    }

    @Test
    fun `boot keeps a stored slice on top of the bundle when it is newer and in the same language`() {
        val transport = FakeTransport()
        val newer = (CURSOR.toLong() + 60_000).toString()
        val storage = storageWith(
            StorageKeys.CURRENT_LANGUAGE to "en",
            StorageKeys.NAMESPACES to "[\"default\"]",
            StorageKeys.TRANSLATIONS to "{\"Bonjour\":\"Hello, reviewed\"}",
            StorageKeys.LAST_REFRESH to newer,
        )
        val client = boot(transport, storage, Loads().bundle)

        assertEquals("Hello, reviewed", client.t("Bonjour"))
        assertEquals("Thanks", client.t("Merci"))
        assertEquals(mapOf("Bonjour" to "Hello, reviewed", "Merci" to "Thanks"), slice(storage))
        assertEquals(newer, storage.getItem(StorageKeys.LAST_REFRESH))
        assertTrue(transport.dictionaries.isEmpty())
    }

    @Test
    fun `boot ignores a stored slice that is older than the bundle`() {
        val transport = FakeTransport()
        val storage = storageWith(
            StorageKeys.CURRENT_LANGUAGE to "en",
            StorageKeys.NAMESPACES to "[\"default\"]",
            StorageKeys.TRANSLATIONS to "{\"Bonjour\":\"Old hello\"}",
            StorageKeys.LAST_REFRESH to (CURSOR.toLong() - 60_000).toString(),
        )
        val client = boot(transport, storage, Loads().bundle)

        assertEquals("Hello", client.t("Bonjour"))
        assertEquals(CURSOR, storage.getItem(StorageKeys.LAST_REFRESH))
    }

    @Test
    fun `boot never mixes in a newer stored slice written in another language`() {
        val transport = FakeTransport()
        // Storage holds English, the app boots in Spanish (skipCurrentLanguageHydration).
        val storage = storageWith(
            StorageKeys.CURRENT_LANGUAGE to "en",
            StorageKeys.NAMESPACES to "[\"default\"]",
            StorageKeys.TRANSLATIONS to "{\"Bonjour\":\"Hello\"}",
            StorageKeys.LAST_REFRESH to (CURSOR.toLong() + 60_000).toString(),
        )
        val client = boot(
            transport,
            storage,
            Loads().bundle,
            LanguagesConfig(Lang.FR, listOf(Lang.FR, Lang.EN, Lang.ES), initWithDefault = Lang.ES, skipCurrentLanguageHydration = true),
        )

        assertEquals(Lang.ES, client.currentLanguage)
        assertEquals("Hola", client.t("Bonjour"))
        assertEquals(mapOf("Bonjour" to "Hola", "Merci" to "Gracias"), slice(storage))
        assertEquals(CURSOR, storage.getItem(StorageKeys.LAST_REFRESH))
    }

    @Test
    fun `boot falls back to the fetch when the loader throws`() {
        val transport = FakeTransport(mapOf("Fetched" to "From the API"))
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "en")
        val client = boot(transport, storage, Loads(failing = true).bundle)

        assertEquals(2, transport.dictionaries.size)
        assertEquals("From the API", client.t("Fetched"))
    }

    @Test
    fun `a language switch seeds the new language from the bundle and never mixes the previous language in`() {
        val transport = FakeTransport(mapOf("Fetched" to "From the API"))
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "en")
        val client = boot(transport, storage, Loads().bundle)
        assertTrue(transport.dictionaries.isEmpty())

        client.setLanguage(Lang.ES).get()
        client.waitForIdle()

        assertEquals("Hola", client.t("Bonjour"))
        assertEquals(mapOf("Bonjour" to "Hola", "Merci" to "Gracias"), slice(storage))
        assertEquals(CURSOR, storage.getItem(StorageKeys.LAST_REFRESH))
        // checkout has no Spanish file: fetched.
        assertEquals(listOf("https://api.test/translate/es?last_refresh=null&namespace=checkout"), transport.dictionaries.map { it.url })
    }

    @Test
    fun `a miss still POSTs, and the delta after it starts from the bundle cursor`() {
        val transport = FakeTransport(mapOf("Fetched" to "From the API"))
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "en")
        val client = boot(transport, storage, Loads().bundle)

        assertEquals("Nouveau", client.t("Nouveau"))
        client.waitForIdle()

        assertEquals("Nouveau", Json.parse(transport.translates.single().body!!).asMap()["key"])
        assertEquals(listOf("https://api.test/translate/en?last_refresh=$CURSOR"), transport.dictionaries.map { it.url })
        assertEquals("From the API", client.t("Fetched"))
        assertEquals("Hello", client.t("Bonjour"))
    }

    @Test
    fun `without a bundle it fetches exactly as before`() {
        val transport = FakeTransport()
        val storage = storageWith(StorageKeys.CURRENT_LANGUAGE to "en")
        boot(transport, storage, null)

        assertEquals(listOf("https://api.test/translate/en?last_refresh=null"), transport.dictionaries.map { it.url })
    }

    @Test
    fun `from reads the manifest now and the dictionaries on demand`(@TempDir dir: File) {
        File(dir, "default").mkdirs()
        File(dir, "manifest.json").writeText(
            Json.stringify(
                mapOf(
                    "primaryLanguage" to "fr",
                    "languages" to listOf("en", "fr"),
                    "exportedAt" to CURSOR,
                    "namespaces" to mapOf("default" to mapOf("lastRefresh" to CURSOR, "languages" to listOf("en", "fr"))),
                ),
            ),
        )
        File(dir, "default/en.json").writeText(Json.stringify(mapOf("Bonjour" to "Hello")))

        val bundle = I18nKeylessBundle.from { path -> File(dir, path).takeIf { it.isFile }?.readText() }
        assertEquals("fr", bundle.manifest.primaryLanguage)
        assertEquals(listOf("en", "fr"), bundle.manifest.namespaces["default"]?.languages)
        assertEquals(listOf("default"), bundleNamespaces(bundle.manifest))
        assertEquals(mapOf("Bonjour" to "Hello"), bundle.load("default", Lang.EN))
        assertEquals(null, bundle.load("default", Lang.FR))
        assertFalse(bundleCovers(bundle.manifest, "default", "es"))
        assertThrows(JsonException::class.java) { I18nKeylessBundle.from { null } }
    }
}
